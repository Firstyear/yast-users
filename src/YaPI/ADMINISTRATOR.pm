package YaPI::ADMINISTRATOR;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;

textdomain ("users");

# ------------------- imported modules
YaST::YCP::Import ("MailAliases");
YaST::YCP::Import ("Users");
# -------------------------------------

our $VERSION		= '1.0.0';
our @CAPABILITIES 	= ('SLES11');
our %TYPEINFO;

=item *
C<$hash Read ();>

Returns the information about system administrator (root).
Currently return hash contains the list of mail aliases.

=cut

BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any" ]];
}
sub Read {

    my $self	= shift;

    my $root_mail	= MailAliases->GetRootAlias ();
    return {} if !defined $root_mail;

    my @root_aliases	= ();
    foreach my $alias (split (/,/, $root_mail)) {
	$alias	=~ s/[ \t]//g;
	push @root_aliases, $alias;
    }

    return {
	"aliases"	=> \@root_aliases
    }
}

=item *
C<$string Write ($argument_hash);>

write the system adminstrator data. Supported keys of the argument hash are:

    "aliases"	=> list of mail aliases
    "password"	=> new password
    
Returns error message on error.

=cut

BEGIN{$TYPEINFO{Write} = ["function",
    "string",
    [ "map", "string", "any" ]];
}
sub Write {

    my $self	= shift;
    my $args	= shift;
    my $ret	= "";

    if ($args->{"aliases"} && ref ($args->{"aliases"}) eq "ARRAY") {

	my $root_mail	= join (", ", @{$args->{"aliases"}});
	if (!MailAliases->SetRootAlias ($root_mail)) {
	    # error popup
	    $ret = __("An error occurred while setting forwarding for root's mail.");
	    return $ret;
	}
    }
    if ($args->{"password"}) {
	Users->SetRootPassword ($args->{"password"});
	Users->WriteRootPassword ();
    }
    return $ret;
}

