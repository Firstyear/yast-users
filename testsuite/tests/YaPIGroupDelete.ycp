/**
 * $Id$
 */

{
    // testedfiles: Users.pm UsersCache.pm USERS.pm

    include "testsuite.ycp";

    import "YaPI::USERS";
    import "Mode";

    map R = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid"		: "hh",
			"type"		: "local",
			"uidnumber"	: 500,
			"gidnumber"	: 100
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local",
			"more_users":	$[
			    "hh":	1,
			]
		    ],
		    1000: $[
			"cn"	:	"gg",
			"gidnumber":	1000,
			"type"	:	"local",
			"userlist":	$[
			    "hh":	1
			]
		    ],
		    1001: $[
			"cn"	:	"gr",
			"gidnumber":	1001,
			"type"	:	"local",
		    ],
		    "by_name": $[
			"users":	100,
			"gg"	:	1000,
			"gr"	:	1001
		    ],
		    "gids": $[
			100:		1,
			1000:		1,
			1001:		1,
		    ],
		    "groupnames": $[
			"users":	1,
			"gg":		1,
			"gr":		1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0
		    ],
		    "by_name": $[
			"root"		: 0
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		],
		"groups": $[
		    0: $[
			"cn":	"root",
			"gidnumber":	0,
			"type"	:	"system"
		    ],
		    "by_name": $[
			"root"	: 0
		    ],
		    "gids": $[
			0	: 1
		    ],
		    "groupnames": $[
			"root"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/tmp/local/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: 0,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map W = $[];
    map E = $[
	"passwd" : $[
	    "init"	: true
	],
    ];

    DUMP ("==========================================================");

    Mode::test = true;

    map<string,any> config_map  = $[];

    DUMP ("============ delete non-existent group: ==================");
    config_map["cn"]	= "ggg";
    string error = (string)
	TEST (``(YaPI::USERS::GroupDelete (config_map)), [R,W,E], 0);

    DUMP ("==========================================================");
    DUMP ("============ delete groups with members: =================");
    config_map["cn"]	= "gg";
    error = (string)
	TEST (``(YaPI::USERS::GroupDelete (config_map)), [R,W,E], 0);
    DUMP (sformat ("========== returned: %1", error));

    DUMP ("==========================================================");
    DUMP ("============ delete groups with more_users: ==============");
    config_map["cn"]	= "users";
    error = (string)
	TEST (``(YaPI::USERS::GroupDelete (config_map)), [R,W,E], 0);
    DUMP (sformat ("========== returned: %1", error));

    DUMP ("============ delete group by name: =======================");

    config_map["cn"]	= "gr";
    TEST (``(YaPI::USERS::GroupDelete (config_map)), [R,W,E], 0);

    DUMP ("============ delete group by gid: =======================");

    config_map	= $[
	"gidnumber"	: "1001"
    ];
    TEST (``(YaPI::USERS::GroupDelete (config_map)), [R,W,E], 0);
}
