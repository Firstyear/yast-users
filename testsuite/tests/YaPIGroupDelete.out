Dump	==========================================================
Dump	============ delete non-existent group: ==================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], 1000:$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], 1001:$["cn":"gr", "gidnumber":1001, "type":"local"], "by_name":$["gg":1000, "gr":1001, "users":100], "gids":$[100:1, 1000:1, 1001:1], "groupnames":$["gg":1, "gr":1, "users":1]]
Read	.passwd.local.groups.by_name $["gg":1000, "gr":1001, "users":100]
Read	.passwd.system.users $[0:$["uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Return	There is no such group.
Dump	==========================================================
Dump	============ delete groups with members: =================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], 1000:$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], 1001:$["cn":"gr", "gidnumber":1001, "type":"local"], "by_name":$["gg":1000, "gr":1001, "users":100], "gids":$[100:1, 1000:1, 1001:1], "groupnames":$["gg":1, "gr":1, "users":1]]
Read	.passwd.local.groups.by_name $["gg":1000, "gr":1001, "users":100]
Read	.passwd.system.users $[0:$["uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Return	You cannot delete this group because
Dump	========== returned: You cannot delete this group because
Dump	there are users in the group.
Dump	Remove these users from the group first.
Dump	==========================================================
Dump	============ delete groups with more_users: ==============
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], 1000:$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], 1001:$["cn":"gr", "gidnumber":1001, "type":"local"], "by_name":$["gg":1000, "gr":1001, "users":100], "gids":$[100:1, 1000:1, 1001:1], "groupnames":$["gg":1, "gr":1, "users":1]]
Read	.passwd.local.groups.by_name $["gg":1000, "gr":1001, "users":100]
Read	.passwd.system.users $[0:$["uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Return	You cannot delete this group because
Dump	========== returned: You cannot delete this group because
Dump	there are users that use this group
Dump	as their default group.
Dump	============ delete group by name: =======================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], 1000:$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], 1001:$["cn":"gr", "gidnumber":1001, "type":"local"], "by_name":$["gg":1000, "gr":1001, "users":100], "gids":$[100:1, 1000:1, 1001:1], "groupnames":$["gg":1, "gr":1, "users":1]]
Read	.passwd.local.groups.by_name $["gg":1000, "gr":1001, "users":100]
Read	.passwd.system.users $[0:$["uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["100":$["cn":"users", "gidnumber":"100", "more_users":$["hh":"1"], "type":"local"], "1000":$["cn":"gg", "gidnumber":"1000", "type":"local", "userlist":$["hh":"1"]], "by_name":$["gg":"1000", "gr":"1001", "users":"100"], "gids":$["100":"1", "1000":"1", "1001":"1"], "groupnames":$["gg":"1", "gr":"1", "users":"1"]], "system":$["0":$["cn":"root", "gidnumber":"0", "type":"system"], "by_name":$["root":"0"], "gids":$["0":"1"], "groupnames":$["root":"1"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Dump	============ delete group by gid: =======================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], 1000:$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], 1001:$["cn":"gr", "gidnumber":1001, "type":"local"], "by_name":$["gg":1000, "gr":1001, "users":100], "gids":$[100:1, 1000:1, 1001:1], "groupnames":$["gg":1, "gr":1, "users":1]]
Read	.passwd.local.groups.by_name $["gg":1000, "gr":1001, "users":100]
Read	.passwd.system.users $[0:$["uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["100":$["cn":"users", "gidnumber":"100", "more_users":$["hh":"1"], "type":"local"], "1000":$["cn":"gg", "gidnumber":"1000", "type":"local", "userlist":$["hh":"1"]], "by_name":$["gg":"1000", "gr":"1001", "users":"100"], "gids":$["100":"1", "1000":"1", "1001":"1"], "groupnames":$["gg":"1", "gr":"1", "users":"1"]], "system":$["0":$["cn":"root", "gidnumber":"0", "type":"system"], "by_name":$["root":"0"], "gids":$["0":"1"], "groupnames":$["root":"1"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
